### DISCUSSION TITLE
Error in WCS Core test: The ExtendedCapabilities section shall be provided in the path

### DISCUSSION THREAD

**Discussion by nmtoken** [INSPIRE-MIF/helpdesk-validator#775] [General]
- [ ] a description of the steps to reproduce your issue

Using a static copy of GetCapabilities response for a non-public service WCS 2.0.1 service

Using the staging-inspire-validator

- [ ] the browser and version used
Opera
Version:60.0.3255.95

Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.103 Safari/537.36 OPR/60.0.3255.95

- [ ] a screenshot illustrating the problem
- [ ] the Test Report (ideally as an attachment to the issue)

http://staging-inspire-validator.eu-west-1.elasticbeanstalk.com/etf-webapp/v2/TestRuns/EID2a1be7ca-76e8-46ef-81e1-23ab778246c2.html?lang=en#EIDe48e116f-016a-1000-ba75-f9ae74149ad9

The issue reported is **The ExtendedCapabilities section shall be provided in the path: /Capabilities/OperationsMetadata/ExtendedCapabilities/ExtendedCapabilities**

The example GetCapabilities document though has 

at **/wcs:Capabilities/wcs:ServiceMetadata[1]/wcs:Extension[1]/ows:ExtendedCapabilities[1]/*[namespace-uri()='http://inspire.ec.europa.eu/schemas/inspire_dls/1.0' and local-name()='ExtendedCapabilities'][1]**  which is schema valid

- [ ] the name of the Test that failed or better: change the `Level of Detail`

- [ ] the URL of your service / the file you have uploaded or referenced (as
an attachment).

http://ogcdev.bgs.ac.uk/WCS/ESexample.xml

**Comment by danielnavarrogeo** [comment]
Dear @nmtoken 

Thank you for your comments.

We do not really understand what is the problem you are facing here.

As you mentioned above, the test fails because the path to the ExtendedCapabilities section differs with the expected one.

You have provided already, in issue 55, a valid example of the location of the ExtendedCapabilities section in: http://ogc2.bgs.ac.uk/cgi-bin/GEMAS/ows?SERVICE=WCS&acceptVERSIONs=2.0.1&REQUEST=GetCapabilities&

Please, could you clarify us what is the error you have found?

Regards

**Comment by nmtoken** [comment]
The issue is that the expected XPath is too strict.

The tested service is schema valid and includes an extended capabilities section.  It is perhaps ironic that it's the principal service I was using when editing the WCS TG, it shouldn't fail the test.

EDIT

The service conforms to the requirements specified in the technical guidance, (where there is no XPath given to the location of the ExtendedCapabilities section, just relative paths within it)






-----Original Message-----

**Comment by danielnavarrogeo** [comment]
Dear @nmtoken 

Thank you for opening this issue.

We have discussed about it and it is agreed to be consistent with WFS. Therefore, the ExtendedCapabilities section shall be provided in exactly this path:

/Capabilities/OperationsMetadata/ExtendedCapabilities/ExtendedCapabilities

The TG will be updated to reflect this (at the moment, this is not included in the WCS TG).

Regards

**Comment by nmtoken** [comment]
I'm not sure I understand the comment `it is agreed to be consistent with WFS`.   Who agreed it?

Are you saying that the XPath you currently have is the same as used by WFS and that it must be used by WCS too, seems over draconian.

**/Capabilities/OperationsMetadata/ExtendedCapabilities/** seems semantically incorrect, as the ExtendedCapabilites section does not define Operations Metadata.

The correct location (wcsAll.xsd) for adding additional information (such as application profiles, and extensions) is wcs:Extension as in the example tested.

As inspire_dls.xsd, doesn't itself define a location for it's inclusion, only the metadata, it seems quite odd to change the TG to force a less semantically correct view.  Pragmatically better to leave the TG as is with respect  relative paths to required metadata defined in inspire_dls.xsd.  Then leave it up to implementers to decide where best to add the extension, as long as it's schema valid.

**Comment by michellutz** [comment]
Dear James,

to clarify - the resolution mentioned by Daniel was discussed with the validator team here at JRC. The argumentation was that the ows:Capabilities schema (http://schemas.opengis.net/ows/2.0/owsGetCapabilities.xsd), on which also the wcs:Capabilities build, already includes an ExtendedCapabilities element in the operations metadata (http://schemas.opengis.net/ows/2.0/owsOperationsMetadata.xsd). This is already used in the WFS implementations, so we propose to use it also in other OWS-based implementations like the WCS, for consistency. 

But probably this should be discussed and confirmed by the 2017.4 sub-group.

**Comment by nmtoken** [comment]
Thanks Michel for the clarification. 

It is is perhaps slightly odd that you are referencing OWS common 2 for WFS, as even WFS 2.0 the interface standard doesn't reference OWS Common 2, but rather uses OWS Common 1.1.0 (OGC 06-121r3:2009).

We could have a wider discussion about semantics sometime, but The ExtendedCapabilites in OperationsMetadata of OWS common 2, is intended  _...to provide metadata about any additional server abilities._  which I take to mean additional operations of a server, like GetStyles, PutStyles, GetTransect, GetTimeseries, GetVerticalProfile.. that are offered by a service but not part of the specification operations.

**Comment by michellutz** [comment]
OK, but the structure of the OperationsMetadata element is the same in OWS Common 1.1.0.

But I agree that we should have a discussion whether this is the right place for including the INSPIRE extended capabilities.

**Comment by alitka** [comment]
As mentioned in 2017.4 meeting 2019-07-16, I forward this issue to some experts in Germany. As soon as I received a feedback, I will post this here.

**Comment by MarcoMinghini** [comment]
**2017.4 meeting 2019-07-17**

- There are valid reasons for both requiring the same path as WFS (consistency with WFS) and allowing a free choice of the path (semantic inconsistency with the purpose of the _ExtendedCapabilites_ section)
- The same discussion came out also when writing the WCS TG.
- It is decided to leave the discussion open and ask for more feedback by national WCS experts.
- [ ] All to ask feedback to WCS national experts and report the feedback as a comment to this issue.

**Comment by nmtoken** [comment]
> There are valid reasons for both requiring the same path as WFS (consistency with WFS) 

The question perhaps arises as to whether the location defined for WFS is actually correct, and if it isn't how will it align to the WCS position.

More relevant perhaps is the adoption of OGC API-Features (WFS 3) this will have a completely different structure, I think, and is another reason why we shouldn't be too prescriptive of where the data is, rather that the data exists and is correct.

**Comment by alitka** [comment]
> 
> 
> As mentioned in 2017.4 meeting 2019-07-16, I forward this issue to some experts in Germany. As soon as I received a feedback, I will post this here.

Feedback from our national experts:

We totally agree with the previous comments, because: 
1. the path to the ExtendedCapabilities is already clearly defined in the OWS 2.0 scheme
2. the uniformity and therefore user-friendliness of download services are enhance

Furthermore from our point of view, there are no conclusive arguments for locating the Extended Capabilities on a different path than it already is in case of WFS.

**Comment by nmtoken** [comment]
It is not in dispute that there is a location for an (as opposed 'the') ExtendedCapabilities section in the OWS schema used by WCS 2.  The issue is that the this ExtendedCapabilities is for Operations Metadata, and the INSPIRE extended capabilities section does **not** provide Operations Metadata.  

WCS 2 provides another location as an extension to the service metadata, to place such ExtendedCapabilities.

I have also been in contact with German experts on WCS and they agree with this point of view

**Comment by MarcoMinghini** [comment]
Dear all,
as agreed in the last sub-group meeting, could you please ask for feedback about this topic to your WCS national experts and report it here? Thank you.

**Comment by reporiikka** [comment]
Dear all,
we in Finnish Environment Institute SYKE have encountered same problems as  @nmtoken validating our INSPIRE WCS (published with GeoServer). We asked the Finnish INSPIRE secretariat about this, and they told that GeoServer INSPIRE WCS extension was made according to technical guidelines and thus should be INSPIRE compatible. Could it be possible to have both Geoserver and the other schema discussed here valid? 

Regards

**Comment by dperezBM** [comment]
Dear all,

Since this issue had no interaction time ago, we decided to close it. Please feel free to open a new one if needed.

Thank you and best regards.

**Comment by nmtoken** [comment]
So feed back was sought and received, what now for a solution to the issue?

**Comment by KathiSchleidt** [comment]
Dear all,

a short note that there has been a successful WCS validation. The extended INSPIRE information is provided as specified in the TG under `ows:OperationsMetadata/ows:ExtendedCapabilities` by the rasdaman implementation. Demo endpoint available at:

https://inspire.rasdaman.org/rasdaman/ows#/services

While formally valid according to the WCS Schema to place this information under `wcs:Contents/wcs:Extension`, this is in contradiction to the TG, validator complains.

**Comment by KathiSchleidt** [comment]
To my view, this thread (as well as #660) can both be closed as resolved. If data providers wish to utilize the `wcs:Contents/wcs:Extension` option, a new issue with a change request (ideally with exemplary implementation) should be provided.

**Comment by fabiovinci** [comment]
Since there is a long discussion here still open, I am converting this issue into a discussion.

If there is agreement on how to proceed, we will open a new issue.

**Reply by joey-geoloog** [reply]
A few years down the line @fabiovinci but have just found this problem myself. Indeed, we are using GeoServer as posted here.
The discussion I can find online hints no wish for GeoServer to develop this as it's valid following WCS spec.

I see earlier in this thread that if we are providing data this way, that we should look to update the test suite.
How would we go about formalising the wish to adjust the test suite to also see wcs:Extension/* as a valid parent element for ExtendedCapabilities in a WCS service?

https://gis.stackexchange.com/questions/415766/modify-replace-getcapabilities-response-geoserver-inspire-extension-wcs

**Reply by joey-geoloog** [reply]
I have checked further and it would indicate that GeoServer is compliant in Scenario 1 (external metadata)
<img width="879" height="180" alt="image" src="https://github.com/user-attachments/assets/0fd9d1b4-fb9c-4b56-aea6-d9a58cbbbfeb" />

As such, the test suite shouldn't fail if external metadata is being used. In this case, it appears that the test suite should be extended to also validate Scenario 1 as correct. What's the best way to get this handled, open a new issue?
